lib/src/main/kotlin/com/swmansion/starknet/data/serializers/JsonRpcTransactionTracePolymorphicSerializer.kt

```diff
@@ -6,15 +6,25 @@ import kotlinx.serialization.json.*
 
 internal object JsonRpcTransactionTracePolymorphicSerializer :
     JsonContentPolymorphicSerializer<TransactionTrace>(TransactionTrace::class) {
+    private fun selectInvokeTransactionTraceDeserializer(jsonObject: JsonObject): DeserializationStrategy<out InvokeTransactionTrace> {
+        val executeInvocation = jsonObject["execute_invocation"]?.jsonObject ?: throw IllegalStateException("execute_invocation is missing.")
+        val isReverted = "revert_reason" in executeInvocation
+
+        return when (isReverted) {
+            true -> RevertedInvokeTransactionTrace.serializer()
+            false -> CommonInvokeTransactionTrace.serializer()
+        }
+    }
+
     override fun selectDeserializer(element: JsonElement): DeserializationStrategy<out TransactionTrace> {
         val jsonObject = element.jsonObject
 
-        return when (jsonObject.keys) {
-            setOf("validate_invocation", "execute_invocation", "fee_transfer_invocation") -> InvokeTransactionTrace.serializer()
-            setOf("validate_invocation", "fee_transfer_invocation") -> DeclareTransactionTrace.serializer()
-            setOf("validate_invocation", "constructor_invocation", "fee_transfer_invocation") -> DeployAccountTransactionTrace.serializer()
-            setOf("function_invocation") -> L1HandlerTransactionTrace.serializer()
-            else -> throw IllegalArgumentException("Invalid transaction trace type")
+        return when {
+            "execute_invocation" in jsonObject -> selectInvokeTransactionTraceDeserializer(jsonObject)
+            "constructor_invocation" in jsonObject -> DeployAccountTransactionTrace.serializer()
+            "function_invocation" in jsonObject -> L1HandlerTransactionTrace.serializer()
+            listOf("validate_invocation", "fee_transfer_invocation").any { it in jsonObject } -> DeclareTransactionTrace.serializer()
```

Comment:
Let's remember to update this when 0.5.0 comes out - they add tx types in responses there, and it should make our life easier

Reply:
I remember, I initially suggested `type` field to be added 😅 
Or do you want me to add a comment and link this to an issue?

Reply:
nah, if you remember about this that's fine for me :)